### PR TITLE
Feature/logging

### PR DESCRIPTION
**Pull request by Shahar-Y**
Added a logger to the requests from api-gateway. 
It works in Kibana.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,14 +43,17 @@
     "express": "^4.16.4",
     "express-session": "^1.16.1",
     "grpc": "^1.21.1",
+    "grpc-middleware": "^1.1.0",
```
Comment (omrishtam) [inline package.json]: Remove unused dependency.

File: src/file/file.rpc.ts [hunk 2]
```diff
@@ -222,3 +232,41 @@ class ResFile {
...
+ */
+function logOnEntry(methodName : string, fields: any) : void {
+  let description : string = 'request parameters: { ';
+  for (const key of Object.keys(fields)) {
+    const fieldName : string = fields[key].toString();
+    description += `${key} : ${fieldName}, `;
+  }
+  description += ' }';
+
+  const traceId : string = apm.currentTransaction.traceparent.split('-')[1];
```
[the remark this answers is not in the document]
Comment (omrishtam) [reply src/file/file.rpc.ts]: Also I think you should get the trace id once in the wrapper and pass it down to `logOnEntry` and the rest of the `logOnXXX`.

**Comment by omerzamir** [comment]
@shahar-Y You got off easy this time...

**Comment by Shahar-Y** [comment]
@omerzamir I wouldn't say that, if we're taking into account to the number of lines I added...

**Comment by omrishtam** [comment]
> @omerzamir I wouldn't say that, if we're taking into account to the number of lines I added...

@Shahar-Y I can request more things if you want, there are more...

File: src/utils/logger.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+logger.add(elasticsearch);
+
+/**
+ * logs the data with its given parameters.
+ * @param severity - the kind of log created.
+ * @param name - name of the log. in our case, the function called.
+ * @param description - description in text.
+ * @param correlationId - id to correlate to if there are several logs with some connection.
+ * @param user - the user requesting for the service.
+ * @param more - additional optional information.
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/utils/logger.ts]: It is in winston...

[the remark this answers is not in the document]
Comment (omrishtam) [reply src/utils/logger.ts]: image
No it isn't.

File: src/file/file.rpc.ts [hunk 4]
```diff
@@ -239,34 +240,38 @@ class ResFile {
...
 /**
  * logs when the method finished without an error.
  * @param methodName - name of the method called.
+ * @param traceID - the current trace's id.
  */
-function logOnFinish(methodName : string) : void {
-  const traceId : string = apm.currentTransaction.traceparent.split('-')[1];
-  log('info', methodName, 'Finished successfully', traceId);
+function logOnFinish(methodName : string, traceID: string, res: any) : void {
+  JSON.stringify(res);
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: WTF

File: src/file/file.rpc.ts [hunk 5]
```diff
@@ -239,34 +240,38 @@ class ResFile {
...
+  log(Severity.INFO, methodName, 'request ended', traceID, res);
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: Change `request ended` to `response`.

File: src/file/file.rpc.ts [hunk 6]
```diff
@@ -239,34 +240,38 @@ class ResFile {
...
+function logOnError(methodName : string, err: Error, traceID: string) : void {
+  log(Severity.ERROR, methodName, err.message, traceID);
+}
+
+function getCurrTraceId() : string {
+  try {
+    return apm.currentTransaction.traceparent.split('-')[1];
+  } catch (err) {
+    // Should never get here. The log is set after apm starts.
+    return 'NoTraceID';
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: return empty string...

File: src/utils/logger.ts [hunk 7]
```diff
@@ -27,10 +27,19 @@ logger.add(elasticsearch);
...
  * @param name - name of the log. in our case, the function called.
  * @param description - description in text.
- * @param correlationId - id to correlate to if there are several logs with some connection.
+ * @param traceID - id to correlate to if there are several logs with some connection.
  * @param user - the user requesting for the service.
  * @param more - additional optional information.
  */
-export const log = (severity: string, name: string, description: string, correlationId?: string, user?: string, more?: any) => {
-  logger.log(severity, { name, description, correlationId, user, ...more });
+export const log = (severity: Severity, name: string, description: string, traceID?: string, user?: string, more?: any) => {
```
Comment (omrishtam) [inline src/utils/logger.ts]: Think about what happens when `more` is a string...

Comment (omrishtam) [reply src/utils/logger.ts]: And other cases.

File: src/file/file.rpc.ts (hunk 5 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: it's not really indicative..

File: src/utils/logger.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (omrishtam) [reply src/utils/logger.ts]: @Shahar-Y Try this:
```
export const log = (level: Severity, name: string, description: string, traceID?: string, user?: string, more?: object) => {
  logger.log({ level, name, description, traceID, user, ...<any>more });
};
```

and look at this PR in winston: https://github.com/winstonjs/winston/issues/1242

File: src/utils/logger.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * logs the data with its given parameters.
+ * @param severity - the kind of log created.
+ * @param name - name of the log. in our case, the function called.
+ * @param description - description in text.
+ * @param traceID - id to correlate to if there are several logs with some connection.
+ * @param user - the user requesting for the service.
+ * @param more - additional optional information.
+ */
+export const log = (severity: Severity, name: string, description: string, traceID?: string, user?: string, more?: any) => {
```
Comment (omrishtam) [inline src/utils/logger.ts]: Remove `user` parameter.

File: src/utils/logger.ts (hunk 7 above)
Comment (omrishtam) [reply src/utils/logger.ts]: Rename `more` to `meta`.

File: src/utils/logger.ts (hunk 8 above)
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

File: src/utils/logger.ts (hunk 7 above)
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

File: src/file/file.rpc.ts (hunk 6 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts (hunk 5 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts (hunk 4 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts [hunk 9]
```diff
@@ -222,3 +232,41 @@ class ResFile {
...
+  log('info', methodName, description, traceId);
+}
+
+/**
+ * logs when the method finished without an error.
+ * @param methodName - name of the method called.
+ */
+function logOnFinish(methodName : string) : void {
+  const traceId : string = apm.currentTransaction.traceparent.split('-')[1];
+  log('info', methodName, 'Finished successfully', traceId);
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done

File: src/file/file.rpc.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts [hunk 10]
```diff
@@ -222,3 +232,41 @@ class ResFile {
...
+  // ******************** LOGGING FUNCTIONS ******************** */
+
+/**
+ * logs the activity before anything is done.
+ * @param methodName - name of the method activated.
+ * @param fields - parameters sent in the request.
+ */
+function logOnEntry(methodName : string, fields: any) : void {
+  let description : string = 'request parameters: { ';
+  for (const key of Object.keys(fields)) {
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts [hunk 11]
```diff
@@ -51,6 +52,7 @@ export class FileServer {
     this.server = new grpc.Server();
     this.addServices();
     this.server.bind(`0.0.0.0:${port}`, grpc.ServerCredentials.createInsecure());
+    log('info', 'server bind', `server listening on port: ${port}`);
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/utils/logger.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

File: src/utils/logger.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * logs the data with its given parameters.
+ * @param severity - the kind of log created.
+ * @param name - name of the log. in our case, the function called.
+ * @param description - description in text.
+ * @param correlationId - id to correlate to if there are several logs with some connection.
+ * @param user - the user requesting for the service.
+ * @param more - additional optional information.
+ */
+export const log = (severity: string, name: string, description: string, correlationId?: string, user?: string, more?: any) => {
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

File: src/utils/logger.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import * as winston from 'winston';
+import * as os from 'os';
+const Elasticsearch = require('winston-elasticsearch');
```
[the remark this answers is not in the document]
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

File: src/file/file.rpc.ts [hunk 14]
```diff
@@ -74,19 +76,28 @@ export class FileServer {
...
+   * wraps all of the service methods, creating the transaction for the apm and the logger.
+   * @param func - the method called.
+   */
   private wrapper (func: Function) :
   (call: grpc.ServerUnaryCall<Object>, callback: grpc.requestCallback<Object>) => Promise<void> {
     return async (call: grpc.ServerUnaryCall<Object>, callback: grpc.requestCallback<Object>) => {
       try {
         const traceparent = call.metadata.get('elastic-apm-traceparent');
-        const transOptions = (traceparent.length > 0) ? { childOf: traceparent[0].toString() } : {};
+        const transOptions =  (traceparent.length > 0) ? { childOf: traceparent[0].toString() } : {};
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: Remove space.

File: src/file/file.rpc.ts [hunk 15]
```diff
@@ -74,19 +76,28 @@ export class FileServer {
...
         apm.startTransaction(`/file.FileService/${func.name}`, 'request', transOptions);
+        const traceID = getCurrTraceId();
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: Define `const traceID = getCurrTraceId();` once outside `try catch` block

File: src/file/file.rpc.ts [hunk 16]
```diff
@@ -222,3 +233,44 @@ class ResFile {
...
+
+/**
+ * logs the activity before anything is done.
+ * @param methodName - name of the method activated.
+ * @param fields - parameters sent in the request.
+ * @param traceID - the current trace's id.
+ */
+function logOnEntry(methodName : string, fields: any, traceID: string) : void {
+  const description = JSON.stringify(fields);
+  log(Severity.INFO, methodName, description, traceID);
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: There's no indication in the log that it's a request log, like there's on a response log.

File: src/file/file.rpc.ts [hunk 17]
```diff
@@ -222,3 +233,44 @@ class ResFile {
...
+}
+
+/**
+ * logs when the method finished without an error.
+ * @param methodName - name of the method called.
+ * @param traceID - the current trace's id.
+ */
+function logOnFinish(methodName : string, traceID: string, res: any) : void {
+  log(Severity.INFO, methodName, 'Response', traceID);
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: Missing `res`, and change to lower case `response`.

File: src/utils/logger.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * logs the data with its given parameters.
+ * @param severity - the kind of log created.
+ * @param name - name of the log. in our case, the function called.
+ * @param description - description in text.
+ * @param traceID - id to correlate to if there are several logs with some connection.
+ * @param user - the user requesting for the service.
+ * @param meta - additional optional information.
+ */
+export const log = (severity: Severity, name: string, description: string, traceID?: string, meta?: object) => {
```
Comment (omrishtam) [inline src/utils/logger.ts]: Now it's without `any`:
```
export const log = (level: Severity, name: string, message: string, traceID?: string, meta?: object) => {
  logger.log({ level, name, message, traceID, ...meta });
};
```

File: src/file/file.rpc.ts (hunk 16 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/file/file.rpc.ts (hunk 17 above)
Comment (Shahar-Y) [reply src/file/file.rpc.ts]: done.

File: src/utils/logger.ts (hunk 18 above)
Comment (Shahar-Y) [reply src/utils/logger.ts]: done.

**Pull request merged**